diesel/src/connection/transaction_manager.rs

```diff
@@ -333,39 +333,37 @@ where
         let transaction_state = Self::get_transaction_state(conn)?;
 
         let rollback_sql = match transaction_state.in_transaction {
-            Some(ref mut in_transaction) => {
-                match in_transaction.transaction_depth.get() {
-                    1 => Cow::Borrowed("ROLLBACK"),
-                    depth_gt1 => {
-                        if in_transaction.top_level_transaction_requires_rollback {
-                            // There's no point in *actually* rolling back this one
-                            // because we won't be able to do anything until top-level
-                            // is rolled back.
-
-                            // To make it easier on the user (that they don't have to really look
-                            // at actual transaction depth and can just rely on the number of
-                            // times they have called begin/commit/rollback) we don't mark the
-                            // transaction manager as out of the savepoints as soon as we
-                            // realize there is that issue, but instead we still decrement here:
-                            in_transaction.transaction_depth = NonZeroU32::new(depth_gt1 - 1)
-                                .expect("Depth was checked to be > 1");
-                            return Ok(());
-                        } else {
-                            Cow::Owned(format!(
-                                "ROLLBACK TO SAVEPOINT diesel_savepoint_{}",
-                                depth_gt1 - 1
-                            ))
-                        }
-                    }
-                }
-            }
+            Some(ref mut in_transaction) => match in_transaction.transaction_depth.get() {
+                1 => Cow::Borrowed("ROLLBACK"),
+                depth_gt1 => Cow::Owned(format!(
+                    "ROLLBACK TO SAVEPOINT diesel_savepoint_{}",
+                    depth_gt1 - 1
+                )),
+            },
             None => return Err(Error::NotInTransaction),
         };
 
         match conn.batch_execute(&rollback_sql) {
             Ok(()) => {
-                Self::get_transaction_state(conn)?
-                    .change_transaction_depth(TransactionDepthChange::DecreaseDepth)?;
+                let tm_status = &mut Self::get_transaction_state(conn)?;
+                if let ValidTransactionManagerStatus {
+                    in_transaction:
+                        Some(InTransactionStatus {
+                            ref mut top_level_transaction_requires_rollback,
+                            ..
+                        }),
+                } = tm_status
+                {
+                    // rolling back the save point did work, even if the connection
+                    // was marked to require a top level rollback. This likely means that
+                    // we recovered the transaction state and we are able to perform
```

Comment:
So I've tested this and actually it's not true :(

Conn1:
```sql
BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE ;
SELECT * FROM users;
SAVEPOINT sp1;
INSERT INTO users(name) VALUES ('u1');
SAVEPOINT sp2;
```

Conn2:
```sql
BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE ;
SELECT * FROM users;
INSERT INTO users(name) VALUES ('u2');
COMMIT;
```

Conn1:
```sql
SELECT * FROM users; -- errors
ROLLBACK TO SAVEPOINT sp2; -- succeeds
SELECT * FROM users; -- still errors
```

Reply:
I think this approach still has potential - we can may be able to live with a connection on which we only know that it "may" require rollback, ignoring otherwise relevant errors and behaving nicely with regards to rollback attempt on commit fail - I don't think we need to know that we *for sure* require rollback.

I'll try to add a commit on that path here.

Also maybe pg gives us information about whether the connection actually was fixed through the transaction_status, which would enable us to be more accurate there.

Reply:
I've pushed your example as (failing) testcase. I need to stop working on that now, can probably continue on Monday.

Reply:
Ahaha fun fact: if hitting "commit" after receiving a serialization failure (so same commands as previously), PG returns that the commit has succeeded, but in fact it was actually rolled back.